File: image_matcher.py
"""
사진(이미지) 기반 매크로 - 화면에서 템플릿 이미지를 찾는 모듈
"""
from __future__ import annotations

import cv2
import numpy as np
from pathlib import Path
from typing import Optional

try:
    import mss
    import mss.tools
except ImportError:
    mss = None

try:
    import pyautogui
except ImportError:
    pyautogui = None


def capture_screen(region: Optional[tuple[int, int, int, int]] = None) -> np.ndarray:
    """
    화면을 캡처하여 BGR numpy 배열로 반환.
    region: (x, y, width, height) 또는 None이면 전체 화면
    """
    if mss:
        with mss.mss() as sct:
            if region:
                x, y, w, h = region
                monitor = {"left": x, "top": y, "width": w, "height": h}
            else:
                monitor = sct.monitors[0]
            img = np.array(sct.grab(monitor))
            return cv2.cvtColor(img, cv2.COLOR_BGRA2BGR)
    if pyautogui:
        if region:
            x, y, w, h = region
            img = pyautogui.screenshot(region=(x, y, w, h))
        else:
            img = pyautogui.screenshot()
        return cv2.cvtColor(np.array(img), cv2.COLOR_RGB2BGR)
    raise RuntimeError("mss 또는 pyautogui가 필요합니다. pip install mss pyautogui")


def find_template(
    screen: np.ndarray,
    template_path: str | Path,
    threshold: float = 0.8,
    method: int = cv2.TM_CCOEFF_NORMED,
) -> list[tuple[int, int]]:
    """
    화면 이미지에서 템플릿 이미지를 찾아 매칭된 좌표 목록 반환.
    반환: [(center_x, center_y), ...] - 템플릿 중심 좌표
    """
    template = cv2.imread(str(template_path))
    if template is None:
        raise FileNotFoundError(f"템플릿 이미지를 찾을 수 없습니다: {template_path}")

    screen_gray = cv2.cvtColor(screen, cv2.COLOR_BGR2GRAY)
    template_gray = cv2.cvtColor(template, cv2.COLOR_BGR2GRAY)
    th, tw = template_gray.shape[:2]

    result = cv2.matchTemplate(screen_gray, template_gray, method)
    if method in (cv2.TM_SQDIFF, cv2.TM_SQDIFF_NORMED):
        locations = np.where(result <= 1 - threshold)
    else:
        locations = np.where(result >= threshold)

    centers = []
    for pt in zip(*locations[::-1]):
        cx = pt[0] + tw // 2
        cy = pt[1] + th // 2
        centers.append((cx, cy))
    return centers


def find_on_screen(
    template_path: str | Path,
    region: Optional[tuple[int, int, int, int]] = None,
    threshold: float = 0.8,
) -> Optional[tuple[int, int]]:
    """
    현재 화면에서 템플릿을 찾아 첫 번째 매칭의 중심 (x, y) 반환.
    없으면 None.
    """
    screen = capture_screen(region)
    matches = find_template(screen, template_path, threshold=threshold)
    if not matches:
        return None
    if region:
        rx, ry, _, _ = region
        mx, my = matches[0]
        return (rx + mx, ry + my)
    return matches[0]


def save_region_as_template(
    region: tuple[int, int, int, int],
    save_path: str | Path,
) -> None:
    """화면의 지정 영역을 캡처해 템플릿 이미지로 저장"""
    img = capture_screen(region)
    cv2.imwrite(str(save_path), img)


def find_on_screen_orb(
    template_path: str | Path,
    region: Optional[tuple[int, int, int, int]] = None,
    min_match_count: int = 8,
) -> Optional[tuple[int, int]]:
    """
    ORB 특징점을 이용해 더 강력하게 템플릿 위치를 찾는다.
    - 크기/밝기/약간의 회전에 더 강함.
    - 매칭이 불충분하면 None 반환.
    """
    template_path = Path(template_path)
    template = cv2.imread(str(template_path), cv2.IMREAD_GRAYSCALE)
    if template is None:
        raise FileNotFoundError(f"템플릿 이미지를 찾을 수 없습니다: {template_path}")

    screen_bgr = capture_screen(region)
    screen_gray = cv2.cvtColor(screen_bgr, cv2.COLOR_BGR2GRAY)

    # ORB 특징점 검출
    orb = cv2.ORB_create(nfeatures=500)
    kp1, des1 = orb.detectAndCompute(template, None)
    kp2, des2 = orb.detectAndCompute(screen_gray, None)
    if des1 is None or des2 is None or len(kp1) == 0 or len(kp2) == 0:
        return None

    # BFMatcher + Lowe ratio test
    bf = cv2.BFMatcher(cv2.NORM_HAMMING, crossCheck=False)
    matches = bf.knnMatch(des1, des2, k=2)

    good = []
    for m, n in matches:
        if m.distance < 0.75 * n.distance:
            good.append(m)

    if len(good) < min_match_count:
        return None

    src_pts = np.float32([kp1[m.queryIdx].pt for m in good]).reshape(-1, 1, 2)
    dst_pts = np.float32([kp2[m.trainIdx].pt for m in good]).reshape(-1, 1, 2)

    # 호모그래피로 위치 추정
    M, mask = cv2.findHomography(src_pts, dst_pts, cv2.RANSAC, 5.0)
    if M is None:
        return None

    h, w = template.shape
    corners = np.float32(
        [[0, 0], [0, h - 1], [w - 1, h - 1], [w - 1, 0]]
    ).reshape(-1, 1, 2)
    dst = cv2.perspectiveTransform(corners, M)

    cx = int(dst[:, 0, 0].mean())
    cy = int(dst[:, 0, 1].mean())

    if region:
        rx, ry, _, _ = region
        return rx + cx, ry + cy
    return cx, cy


def _crop_whitespace(img: np.ndarray, bg_thresh: int = 240) -> np.ndarray:
    """흰색/밝은 여백을 자동으로 잘라냄"""
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    # 밝은 배경이 아닌 부분 찾기
    mask = gray < bg_thresh
    coords = cv2.findNonZero(mask.astype(np.uint8))
    if coords is None:
        return img
    x, y, w, h = cv2.boundingRect(coords)
    # 약간의 패딩 추가
    pad = 2
    x = max(0, x - pad)
    y = max(0, y - pad)
    w = min(img.shape[1] - x, w + pad * 2)
    h = min(img.shape[0] - y, h + pad * 2)
    return img[y : y + h, x : x + w]


def find_template_multiscale(
    screen: np.ndarray,
    template_path: str | Path,
    threshold: float = 0.75,
    scales: list[float] | None = None,
) -> Optional[tuple[int, int]]:
    """
    멀티스케일 + 크롭된 템플릿으로 매칭.
    - 템플릿 크기가 실제 화면 아이콘과 조금 달라도 더 잘 찾도록 보완.
    """
    template = cv2.imread(str(template_path))
    if template is None:
        raise FileNotFoundError(f"템플릿: {template_path}")

    # ① 흰 여백 자동 크롭
    template = _crop_whitespace(template)

    screen_gray = cv2.cvtColor(screen, cv2.COLOR_BGR2GRAY)
    template_gray = cv2.cvtColor(template, cv2.COLOR_BGR2GRAY)

    if scales is None:
        scales = [0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2, 1.5, 2.0]

    best_val = -1.0
    best_loc: Optional[tuple[int, int]] = None
    best_tw, best_th = 0, 0

    for scale in scales:
        tw = int(template_gray.shape[1] * scale)
        th = int(template_gray.shape[0] * scale)
        if (
            tw < 10
            or th < 10
            or tw > screen_gray.shape[1]
            or th > screen_gray.shape[0]
        ):
            continue

        resized = cv2.resize(template_gray, (tw, th), interpolation=cv2.INTER_AREA)
        result = cv2.matchTemplate(screen_gray, resized, cv2.TM_CCOEFF_NORMED)
        _, max_val, _, max_loc = cv2.minMaxLoc(result)

        if max_val > best_val:
            best_val = max_val
            best_loc = max_loc
            best_tw, best_th = tw, th

    if best_loc is not None and best_val >= threshold:
        cx = best_loc[0] + best_tw // 2
        cy = best_loc[1] + best_th // 2
        return (cx, cy)
    return None


def find_on_screen_multiscale(
    template_path: str | Path,
    region: Optional[tuple[int, int, int, int]] = None,
    threshold: float = 0.75,
    scales: list[float] | None = None,
) -> Optional[tuple[int, int]]:
    """화면 전체(or 영역)에서 멀티스케일 템플릿 매칭으로 중심 좌표를 찾는다."""
    screen = capture_screen(region)
    pos = find_template_multiscale(screen, template_path, threshold=threshold, scales=scales)
    if pos is None:
        return None
    cx, cy = pos
    if region:
        rx, ry, _, _ = region
        return rx + cx, ry + cy
    return cx, cy

File: MacroWinForms/ImageMatcher.cs
using OpenCvSharp;
using System.Drawing;
using System.Drawing.Imaging;

namespace MacroWinForms;

/// <summary>
/// 화면 캡처 및 템플릿 이미지 매칭 (Python image_matcher.py 대응)
/// </summary>
public static class ImageMatcher
{
    /// <summary>
    /// 화면을 캡처하여 BGR Mat으로 반환.
    /// region: (x, y, width, height) 또는 null이면 전체 화면
    /// </summary>
    public static Mat CaptureScreen(Rectangle? region = null)
    {
        Rectangle bounds = region ?? new Rectangle(
            SystemInformation.VirtualScreen.X,
            SystemInformation.VirtualScreen.Y,
            SystemInformation.VirtualScreen.Width,
            SystemInformation.VirtualScreen.Height);

        using var bmp = new Bitmap(bounds.Width, bounds.Height, PixelFormat.Format32bppArgb);
        using (var g = Graphics.FromImage(bmp))
        {
            g.CopyFromScreen(bounds.Location, System.Drawing.Point.Empty, bounds.Size, CopyPixelOperation.SourceCopy);
        }

        var mat = OpenCvSharp.Extensions.BitmapConverter.ToMat(bmp);
        // Bitmap은 ARGB, OpenCV는 BGR 사용
        Cv2.CvtColor(mat, mat, ColorConversionCodes.BGRA2BGR);
        return mat;
    }

    /// <summary>
    /// 화면 이미지에서 템플릿을 찾아 매칭된 중심 좌표 목록 반환
    /// </summary>
    public static List<(int X, int Y)> FindTemplate(Mat screen, string templatePath, double threshold = 0.8)
    {
        using var template = Cv2.ImRead(templatePath);
        if (template.Empty())
            throw new FileNotFoundException($"템플릿 이미지를 찾을 수 없습니다: {templatePath}");

        using var screenGray = new Mat();
        using var templateGray = new Mat();
        Cv2.CvtColor(screen, screenGray, ColorConversionCodes.BGR2GRAY);
        Cv2.CvtColor(template, templateGray, ColorConversionCodes.BGR2GRAY);

        int tw = templateGray.Width, th = templateGray.Height;
        using var result = new Mat();
        Cv2.MatchTemplate(screenGray, templateGray, result, TemplateMatchModes.CCoeffNormed);

        Cv2.MinMaxLoc(result, out _, out double maxVal, out _, out OpenCvSharp.Point maxLoc);
        var centers = new List<(int, int)>();

        // 첫 번째(최고) 매칭만 반환 (Python find_on_screen은 matches[0]만 사용)
        if (maxVal >= threshold)
        {
            int cx = maxLoc.X + tw / 2;
            int cy = maxLoc.Y + th / 2;
            centers.Add((cx, cy));
        }

        return centers;
    }

    /// <summary>
    /// 현재 화면에서 템플릿을 찾아 첫 번째 매칭의 중심 (x, y) 반환. 없으면 null
    /// </summary>
    public static (int X, int Y)? FindOnScreen(string templatePath, Rectangle? region = null, double threshold = 0.8)
    {
        using var screen = CaptureScreen(region);
        var matches = FindTemplate(screen, templatePath, threshold);
        if (matches.Count == 0) return null;

        var (mx, my) = matches[0];
        if (region.HasValue)
        {
            var r = region.Value;
            return (r.X + mx, r.Y + my);
        }
        return (mx, my);
    }

    /// <summary>
    /// 흰 여백 크롭
    /// </summary>
    private static Mat CropWhitespace(Mat img, byte bgThresh = 240)
    {
        using var gray = new Mat();
        Cv2.CvtColor(img, gray, ColorConversionCodes.BGR2GRAY);
        Cv2.Threshold(gray, gray, bgThresh, 255, ThresholdTypes.BinaryInv);
        using var nonzero = new Mat();
        Cv2.FindNonZero(gray, nonzero);
        if (nonzero.Empty()) return img.Clone();

        var rect = Cv2.BoundingRect(nonzero);
        int pad = 2;
        int x = Math.Max(0, rect.X - pad);
        int y = Math.Max(0, rect.Y - pad);
        int w = Math.Min(img.Width - x, rect.Width + pad * 2);
        int h = Math.Min(img.Height - y, rect.Height + pad * 2);
        return new Mat(img, new OpenCvSharp.Rect(x, y, w, h)).Clone();
    }

    private static readonly double[] DefaultScales = { 0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2, 1.5, 2.0 };

    /// <summary>
    /// 멀티스케일 템플릿 매칭
    /// </summary>
    public static (int X, int Y)? FindOnScreenMultiscale(string templatePath, Rectangle? region = null, double threshold = 0.75, double[]? scales = null)
    {
        scales ??= DefaultScales;
        using var screen = CaptureScreen(region);
        using var templateFull = Cv2.ImRead(templatePath);
        if (templateFull.Empty()) throw new FileNotFoundException($"템플릿: {templatePath}");

        using var template = CropWhitespace(templateFull);
        using var screenGray = new Mat();
        using var templateGray = new Mat();
        Cv2.CvtColor(screen, screenGray, ColorConversionCodes.BGR2GRAY);
        Cv2.CvtColor(template, templateGray, ColorConversionCodes.BGR2GRAY);

        double bestVal = -1.0;
        OpenCvSharp.Point? bestLoc = null;
        int bestTw = 0, bestTh = 0;

        foreach (double scale in scales)
        {
            int tw = (int)(templateGray.Width * scale);
            int th = (int)(templateGray.Height * scale);
            if (tw < 10 || th < 10 || tw > screenGray.Width || th > screenGray.Height) continue;

            using var resized = new Mat();
            Cv2.Resize(templateGray, resized, new OpenCvSharp.Size(tw, th), 0, 0, InterpolationFlags.Area);
            using var result = new Mat();
            Cv2.MatchTemplate(screenGray, resized, result, TemplateMatchModes.CCoeffNormed);
            Cv2.MinMaxLoc(result, out _, out double maxVal, out _, out OpenCvSharp.Point maxLoc);

            if (maxVal > bestVal)
            {
                bestVal = maxVal;
                bestLoc = maxLoc;
                bestTw = tw;
                bestTh = th;
            }
        }

        if (bestLoc.HasValue && bestVal >= threshold)
        {
            int cx = bestLoc.Value.X + bestTw / 2;
            int cy = bestLoc.Value.Y + bestTh / 2;
            if (region.HasValue)
            {
                var r = region.Value;
                return (r.X + cx, r.Y + cy);
            }
            return (cx, cy);
        }
        return null;
    }

    /// <summary>
    /// ORB 특징점으로 템플릿 위치 검출
    /// </summary>
    public static (int X, int Y)? FindOnScreenOrb(string templatePath, Rectangle? region = null, int minMatchCount = 8)
    {
        using var template = Cv2.ImRead(templatePath, ImreadModes.Grayscale);
        if (template.Empty()) throw new FileNotFoundException($"템플릿: {templatePath}");

        using var screenBgr = CaptureScreen(region);
        using var screenGray = new Mat();
        Cv2.CvtColor(screenBgr, screenGray, ColorConversionCodes.BGR2GRAY);

        using var orb = ORB.Create(500);
        using var des1 = new Mat();
        using var des2 = new Mat();
        orb.DetectAndCompute(template, null, out var kp1, des1);
        orb.DetectAndCompute(screenGray, null, out var kp2, des2);
        if (des1.Empty() || des2.Empty() || kp1.Length == 0 || kp2.Length == 0) return null;

        using var bf = new BFMatcher(NormTypes.Hamming, false);
        var matches = bf.KnnMatch(des1, des2, 2);
        var good = matches.Where(m => m[0].Distance < 0.75 * m[1].Distance).Select(m => m[0]).ToArray();
        if (good.Length < minMatchCount) return null;

        var srcPts = good.Select(m => kp1[m.QueryIdx].Pt).Select(p => new Point2f(p.X, p.Y)).ToArray();
        var dstPts = good.Select(m => kp2[m.TrainIdx].Pt).Select(p => new Point2f(p.X, p.Y)).ToArray();
        using var srcMat = InputArray.Create(srcPts);
        using var dstMat = InputArray.Create(dstPts);
        using var M = Cv2.FindHomography(srcMat, dstMat, HomographyMethods.Ransac, 5.0);
        if (M == null || M.Empty()) return null;

        int h = template.Height, w = template.Width;
        var corners = new[] { new Point2f(0, 0), new Point2f(0, h - 1), new Point2f(w - 1, h - 1), new Point2f(w - 1, 0) };
        using var cornersMat = InputArray.Create(corners);
        using var dst = new Mat();
        Cv2.PerspectiveTransform(cornersMat, M, dst);
        double cx = 0, cy = 0;
        for (int i = 0; i < 4; i++)
        {
            var pt = dst.At<Point2f>(i, 0);
            cx += pt.X;
            cy += pt.Y;
        }
        int cxi = (int)(cx / 4), cyi = (int)(cy / 4);

        if (region.HasValue)
            return (region.Value.X + cxi, region.Value.Y + cyi);
        return (cxi, cyi);
    }

    /// <summary>
    /// 화면 영역 캡처하여 Bitmap 반환 (OCR 등에 사용)
    /// </summary>
    public static Bitmap CaptureScreenToBitmap(Rectangle? region = null)
    {
        Rectangle bounds = region ?? new Rectangle(
            SystemInformation.VirtualScreen.X, SystemInformation.VirtualScreen.Y,
            SystemInformation.VirtualScreen.Width, SystemInformation.VirtualScreen.Height);
        var bmp = new Bitmap(bounds.Width, bounds.Height, PixelFormat.Format32bppArgb);
        using (var g = Graphics.FromImage(bmp))
            g.CopyFromScreen(bounds.Location, System.Drawing.Point.Empty, bounds.Size, CopyPixelOperation.SourceCopy);
        return bmp;
    }
}
